File: internal/report/serve.go
package report

import (
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"os/exec"
	"runtime"
)

const host = "127.0.0.1"

func Serve(p Payload, port int, openBrowser bool) error {
	htmlBody := []byte(Render(p))
	raw, _ := json.MarshalIndent(p, "", "  ")

	mux := http.NewServeMux()
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/" && r.URL.Path != "/index.html" {
			http.NotFound(w, r)
			return
		}
		w.Header().Set("Content-Type", "text/html; charset=utf-8")
		w.Header().Set("Cache-Control", "no-store")
		_, _ = w.Write(htmlBody)
	})
	mux.HandleFunc("/report.json", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		w.Header().Set("Cache-Control", "no-store")
		_, _ = w.Write(raw)
	})

	ln, err := net.Listen("tcp", fmt.Sprintf("%s:%d", host, port))
	if err != nil {
		return err
	}
	url := fmt.Sprintf("http://%s", ln.Addr().String())
	fmt.Printf("SecSentry dashboard  %s\n", url)
	fmt.Println("Loopback only. Secrets are masked. Ctrl-C to stop.")
	if openBrowser {
		_ = openURL(url)
	}
	return http.Serve(ln, mux)
}

func openURL(url string) error {
	var cmd *exec.Cmd
	switch runtime.GOOS {
	case "darwin":
		cmd = exec.Command("open", url)
	case "windows":
		cmd = exec.Command("cmd", "/c", "start", url)
	default:
		cmd = exec.Command("xdg-open", url)
	}
	return cmd.Start()
}

File: src/secsentry/reports/serve.py
"""Serve the dashboard on loopback only.

Binds 127.0.0.1 by design — a scan report is a map of where your credentials
leaked, so it must never be reachable from the network. ADR-006.
"""

from __future__ import annotations

import json
import webbrowser
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

from secsentry.reports.html import render

HOST = "127.0.0.1"


def _handler(payload: dict):
    body = render(payload).encode("utf-8")
    raw = json.dumps(payload, indent=2).encode("utf-8")

    class Handler(BaseHTTPRequestHandler):
        def _send(self, content: bytes, content_type: str) -> None:
            self.send_response(200)
            self.send_header("Content-Type", content_type)
            self.send_header("Content-Length", str(len(content)))
            self.send_header("Cache-Control", "no-store")
            self.end_headers()
            self.wfile.write(content)

        def do_GET(self) -> None:  # noqa: N802 - stdlib naming
            if self.path.startswith("/report.json"):
                self._send(raw, "application/json; charset=utf-8")
            elif self.path in ("/", "/index.html"):
                self._send(body, "text/html; charset=utf-8")
            else:
                self.send_error(404)

        def log_message(self, *args) -> None:
            pass

    return Handler


def serve(payload: dict, port: int = 8765, open_browser: bool = True) -> None:
    server = ThreadingHTTPServer((HOST, port), _handler(payload))
    url = f"http://{HOST}:{server.server_port}"
    print(f"SecSentry dashboard  {url}")
    print("Loopback only. Secrets are masked. Ctrl-C to stop.")
    if open_browser:
        try:
            webbrowser.open(url)
        except Exception:
            pass
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nStopped.")
    finally:
        server.server_close()
